src/plugins/matrixops/clip.cpp

```diff
@@ -0,0 +1,182 @@
+
+#include <phylanx/config.hpp>
+#include <phylanx/execution_tree/primitives/node_data_helpers.hpp>
+#include <phylanx/ir/node_data.hpp>
+#include <phylanx/plugins/matrixops/clip.hpp>
+#include <phylanx/util/detail/numeric_limits_min.hpp>
+#include <phylanx/util/matrix_iterators.hpp>
+
+#include <hpx/include/lcos.hpp>
+#include <hpx/include/naming.hpp>
+#include <hpx/include/util.hpp>
+#include <hpx/throw_exception.hpp>
+#include <hpx/util/iterator_facade.hpp>
+
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <blaze/Math.h>
+
+///////////////////////////////////////////////////////////////////////////////
+namespace phylanx { namespace execution_tree { namespace primitives
+{
+    ///////////////////////////////////////////////////////////////////////////
+    match_pattern_type const clip::match_data = {hpx::util::make_tuple("clip",
+        std::vector<std::string>{"clip(_1, _2, _3)"}, &create_clip,
+        &create_primitive<clip>, R"(
+            a, axis
+            Args:
+
+                a (array_like) : array containing elements to clip
+                a_min : Minimum value, it may be scalar or array_like or None
+                a_max : Maximum valus, it may be scalar or array_like or None
+
+            Returns:
+
+            An array with the elements of a, but where values < a_min are replaced with
+            a_min, and those > a_max with a_max."
+            )")};
+
+    ///////////////////////////////////////////////////////////////////////////
+    clip::clip(primitive_arguments_type&& operands, std::string const& name,
+        std::string const& codename)
+      : primitive_component_base(std::move(operands), name, codename)
+    {
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+
+    template <typename T>
+    primitive_argument_type clip::clip0d(ir::node_data<T>&& arg,
+        ir::node_data<T>&& min, ir::node_data<T>&& max) const
+    {
+        auto result = (blaze::max)(
+            min.scalar(), (blaze::min)(max.scalar(), arg.scalar()));
+
+        return primitive_argument_type{ir::node_data<T>{result}};
+    }
+
+    template <typename T>
+    primitive_argument_type clip::clip1d(ir::node_data<T>&& arg,
+        ir::node_data<T>&& min, ir::node_data<T>&& max) const
+    {
+        auto result = (blaze::max)(
+            min.vector(), (blaze::min)(max.vector(), arg.vector()));
+
+        return primitive_argument_type{ir::node_data<T>{result}};
+    }
+
+    template <typename T>
+    primitive_argument_type clip::clip2d(ir::node_data<T>&& arg,
+        ir::node_data<T>&& min, ir::node_data<T>&& max) const
+    {
+        auto result = (blaze::max)(
+            min.matrix(), (blaze::min)(max.matrix(), arg.matrix()));
+
+        return primitive_argument_type{ir::node_data<T>{result}};
+    }
+
+    template <typename T>
```

Comment:
It should be almost trivial to add 3D support here, something like:
```
#if defined(PHYLANX_HAVE_BLAZE_TENSOR)
    template <typename T> 
    primitive_argument_type clip::clip3d(ir::node_data<T>&& arg, 
        ir::node_data<T>&& min, ir::node_data<T>&& max) const 
    { 
        auto result = (blaze::max)( 
            min.tensor(), (blaze::min)(max.tensor(), arg.tensor())); 

        return primitive_argument_type{ir::node_data<T>{result}}; 
    } 
#endif
```
should do the trick.

Reply:
You probably want to call this function as well...